extensions/vertx/runtime/src/main/java/io/quarkus/vertx/runtime/VertxCurrentContextFactory.java

```diff
@@ -41,20 +55,49 @@ public <T extends InjectableContext.ContextState> CurrentContext<T> create(Class
         return new VertxCurrentContext<>(key);
     }
 
-    /**
-     *
-     * @return an unmodifiable list of used keys
-     */
-    public List<String> keys() {
-        return unmodifiableKeys;
+    public ContextInternal duplicateContextIfContainsAnyCreatedScopeKeys(ContextInternal vertxContext) {
+        if (containsAnyCreatedScopeKeys(vertxContext)) {
+            // Duplicate the context, copy the data, remove the request context
+            vertxContext = vertxContext.duplicate();
```

Comment:
I believe this is a bug: you're reassigning the `vertxContext` variable, but if the intent of the subsequent line is also copying the `localContextData` from the old to the new context, this of course won't work.

Reply:
Thanks @mariofusco , yep I've optimized without thinking here; this now should be fine-ish (check the new TODO on the code on why it is not!)